glib/tests/clone.rs

```diff
@@ -416,37 +416,64 @@ fn test_clone_macro_double_rename() {
 
 #[test]
 fn test_clone_macro_typed_args() {
-    let v = Rc::new(1);
-    let w = Rc::new(2);
+    let v = Rc::new(RefCell::new(1));
+    let w = Rc::new(RefCell::new(1));
 
-    let _closure = clone!(@weak v as x, @weak w => @default-panic, move |_x: i8| {
-        println!("v: {}, w: {}", x, w);
+    let closure = clone!(@weak v as x, @weak w => @default-panic, move |arg: i8| {
+        *x.borrow_mut() = arg;
+        *w.borrow_mut() = arg;
     });
+    closure(0);
+    assert_eq!(0, *v.borrow());
+    assert_eq!(0, *w.borrow());
 
-    let _closure = clone!(@weak v, @weak w as x => @default-panic, move |_x: i8| {
-        println!("v: {}, w: {}", v, x);
+    let closure = clone!(@weak v, @weak w as x => @default-panic, move |arg: i8| {
+        *v.borrow_mut() = arg;
+        *x.borrow_mut() = arg;
     });
+    closure(2);
+    assert_eq!(2, *v.borrow());
+    assert_eq!(2, *w.borrow());
 
-    let _closure = clone!(@strong v as x, @strong w => @default-panic, move |_x: i8| {
-        println!("v: {}, w: {}", x, w);
+    let closure = clone!(@strong v as x, @strong w => @default-panic, move |arg: i8| {
+        *x.borrow_mut() = arg;
+        *w.borrow_mut() += arg;
     });
+    closure(1);
+    assert_eq!(1, *v.borrow());
+    assert_eq!(3, *w.borrow());
 
-    let _closure = clone!(@strong v, @strong w as x => @default-panic, move |_x: i8| {
-        println!("v: {}, w: {}", v, x);
+    let closure = clone!(@strong v, @strong w as x => @default-panic, move |arg: i8| {
+        *v.borrow_mut() = arg;
+        *x.borrow_mut() = arg;
     });
+    closure(6);
+    assert_eq!(6, *v.borrow());
+    assert_eq!(6, *w.borrow());
 
-    let _closure = clone!(@weak v as x, @weak w => move |_x: i8| {
-        println!("v: {}, w: {}", x, w);
+    let closure = clone!(@weak v as x, @weak w => move |arg: i8| {
+        *x.borrow_mut() = arg;
+        *w.borrow_mut() = arg;
     });
+    closure(4);
+    assert_eq!(4, *v.borrow());
+    assert_eq!(4, *w.borrow());
 
-    let _closure = clone!(@weak v, @weak w as x => move |_x: i8| {
-        println!("v: {}, w: {}", v, x);
+    let closure = clone!(@weak v, @weak w as x => move |arg: i8| {
+        *v.borrow_mut() = arg;
+        *x.borrow_mut() = arg;
     });
+    closure(1);
+    assert_eq!(1, *v.borrow());
+    assert_eq!(1, *w.borrow());
 
-    let closure = clone!(@weak v, @weak w as x => move |_: i8, _| {
-        println!("v: {}, w: {}", v, x);
+    let closure = clone!(@weak v, @weak w as x => move |arg: i8, arg2| {
+        *v.borrow_mut() = arg;
+        *x.borrow_mut() = arg2;
     });
-    closure(0, 'a');
+    closure(0, 9);
+    assert_eq!(0, *v.borrow());
+    assert_eq!(9, *w.borrow());
```

Comment:
Do we have tests somewhere that check dropping of the weak references so that upgrading fails?

Reply:
Like [here](https://github.com/gtk-rs/gtk-rs/pull/373/files#diff-78ad6fe523da0fb4f9437f0d29ff5d476075b5ff590762b72cdbf13f229fa92bR460) you mean? I can add some to ensure that nothing was done on the values (that's actually a good idea).

Reply:
Yes, sounds good